### DISCUSSION TITLE
Transformed code uses var type instead of dynamic type specified in the original source code..

### DISCUSSION THREAD

**Discussion by HowardShank** [metalama/Metalama#262] [Q&A] [answered]
Hello,

I have the following line of code in my aspect to execute my code:
```
dynamic? result = meta.Proceed();
```
The generated/transformed output code becomes this:
```
var result_1 = await this.ReadAll_Source();
```
The orignal method signature is this:
```
public async Task<ActionResult<IEnumerable<TDtoEntity>>> ReadAll()
```

My OverrideMethod signature and code looks like this:
```
[Template]
private dynamic? OverrideMethod(IFieldOrProperty loggerFieldOrProperty,
            IFieldOrProperty eventLoggerFieldOrProperty,
            IFieldOrProperty auditorFieldOrProperty,
            IFieldOrProperty httpContextFieldOrProperty,
            IFieldOrProperty serviceProviderFieldOrProperty,
            IFieldOrProperty schemaFieldOrProperty)
{
     //Do a bunch of stuff
...
     // Call my method
     dynamic? result = meta.Proceed();
    // Working with the dynamic result (ActionResult, OkObjectResult, IEnumerable objects)
     foreach (var item in result.Result.Value)
    { 
         sb.AppendLine(System.Text.Json.JsonSerializer.Serialize(item, new JsonSerializerOptions { WriteIndented = true }));
    }
...
    // Do more stuff
    return result;
}
```

The problem is, the transformed code changes my dynamic variable to a var variable and the dynamic data is not recognized then.

I've tried Convert.ChangeType(..), Casting, Assigning to individual variables as var & dynamic with no luck.

Once the result comes back as a var I lose the dynamic access to child objects.

Am I doing something wrong or could this be a bug in the transformation?

Thank you!
Howard

**Comment by gfraiteur** [comment]
First, you can try calling `meta.ProceedAsync`, which returns a `Task<dynamic>`.
If you need to cast, then you will also need a [generic template method](https://doc.postsharp.net/metalama/conceptual/aspects/templates/template-parameters) with the type parameter `TDtoEntity`.

**Comment by svick** [comment]
If you know you have `OkObjectResult`, you could cast to that type:

```c#
foreach (var item in (IEnumerable<object>)((OkObjectResult)result.Result).Value)
{
    sb.AppendLine(System.Text.Json.JsonSerializer.Serialize(item, new JsonSerializerOptions { WriteIndented = true }));
}
```

If you actually want to use `dynamic` in your transformed code, then you can't do it directly inside an aspect (because Metalama reserves `dynamic` for its own purposes). But you could call a method in a separate class that uses dynamic:

```c#
ActionResultHelper.AppendResultValues(sb, result);

...

class ActionResultHelper
{
    public static void AppendResultValues(StringBuilder sb, dynamic result)
    {
        foreach (var item in result.Result.Value)
        {
            sb.AppendLine(System.Text.Json.JsonSerializer.Serialize(item, new JsonSerializerOptions { WriteIndented = true }));
        }
    }
}
```

**Reply by HowardShank** [reply]
I tried this and other versions. Becuase var result is not dynamic, **result.Result** is not available in the object. I will try the help class and see if that works out.
```
(IEnumerable<object>)((OkObjectResult)result.Result).Value
```

**Reply by svick** [reply]
In your transformed code, `result` should be `ActionResult<IEnumerable<TDtoEntity>>`, so `result.Result` should be available (but `result.Result.Value` won't, hence the cast to `OkObjectResult`).

**Reply by HowardShank** [reply]
It should, but it doesn't. I either end up with a metalama error if I satisfy the compiler requirements to do this, or I end up with a compiler error if I satisfy the metalama requirements.

**Comment by HowardShank** [comment, marked as answer]
We are going to table this for the time being and just take a standard object oriented approach to add this to the 9 methods that I need it on for right now.

Thanks for the information.

**Reply by gfraiteur** [reply]
Indeed - for just 9 methods a complex aspect does not pay off.

**Comment by benoitc** [comment]
The current behavior is correct per the [ASGI Lifespan Protocol specification](https://asgi.readthedocs.io/en/latest/specs/lifespan.html):

> "Lifespans should be executed once per event loop that will be processing requests. In a multi-process environment there will be lifespan events in each process."

With 4 workers, you get 4 separate processes, each with its own event loop and application instance. Running lifespan once per worker ensures resources like database connections are properly initialized in each process.

For operations that should run only once globally, use gunicorn's server-level hooks:

```python
# gunicorn.conf.py
def when_ready(server):
    print("Runs once in master, before workers spawn")
```
